stan-dev/pystan
# Issue #427: Integrate Betancourt's stan_utility functions into PyStan

#### Summary:

Addition of diagnostic functions to PyStan, based on Michael Betancourt's [stan_utility.py](https://github.com/betanalpha/jupyter_case_studies/blob/master/pystan_workflow/stan_utility.py).

Also, in emulation of similar functionality in RStan, said diagnostics are automatically run after executing the `sampling()` method of `StanModel`.

#### Intended Effect:

To make it harder for users of PyStan to accidentally proceed with a poor sampling of a model.

#### How to Verify:

Currently, the least bad way to verify the diagnostics is to run them on fits of known bad models, such as the centered eight-schools example in http://mc-stan.org/users/documentation/case-studies/pystan_workflow.html

#### Side Effects:

Extra diagnostic information may be printed out.

#### Documentation:

Currently, the documentation is limited to the docstrings of the diagnostic functions.

#### Copyright and Licensing

The diagnostic functions are based on source code from Michael Betancourt, and the license for them as follows:

> The code in this repository is copyrighted by Columbia University and licensed under the new BSD (3-clause) license:  https://opensource.org/licenses/BSD-3-Clause
> 
> The text in this repository is copyrighted by Michael Betancourt and licensed under the CC BY-NC 4.0 license: https://creativecommons.org/licenses/by-nc/4.0/

I gather that my modifications to that code, as well as my modifications to other parts of PyStan, should be licensed as GPL 3.0.
